### PR TITLE
[core] allow offline region metadata to be updated

### PR DESCRIPTION
**Pull request by mollymerp**
WIP #4344 

cc @tobrun

### CODE REVIEW

File: platform/default/mbgl/storage/offline_database.cpp [hunk 1]
```diff
@@ -523,6 +523,19 @@ OfflineRegion OfflineDatabase::createRegion(const OfflineRegionDefinition& defin
...
+    // clang-format off
+    Statement stmt = getStatement(
+                                  "UPDATE regions SET description = ?1"
+                                  "WHERE id = ?2");
+    // clang-format on
+    stmt->bindBlob(1, metadata);
+    stmt->bind(2, region.getID());
+    stmt->run();
+    
+    return metadata;
```
Comment (1ec5) [inline platform/default/mbgl/storage/offline_database.cpp]: Does this method need to return the metadata? Under what circumstances would it differ from the metadata that is passed in?

Comment (mollymerp) [inline platform/default/mbgl/storage/offline_database.cpp]: It doesn't -- I just put it there to keep an error from popping up but it won't be in the final version of this PR 😏

**Comment by tobrun** [comment]
I identified that `OfflineRegion` is the reason @mollymerp and I were having trouble get the code to compile. The offline region is `Move-only` and therefor can only be used as an rValue and can only be constructed from the friend class `OfflineDatabase`. 

I tried optimising the code for this 15ae7df but I'm now hitting an template error in thread.cpp: 

> /Users/Nurbot/work/src/mbgl/util/thread.hpp:68:34: Non-const lvalue reference to type 'vector<[2 * ...]>' cannot bind to a temporary of type 'vector<[2 * ...]>'

Which indicates that I'm not complying to the requirements..
Not sure what I'm doing wrong here.

cc @jfirebaugh @tmpsantos

**Comment by jfirebaugh** [comment]
Hmm, tricky, the API was kind of designed around fully immutable regions, so this is definitely awkward. Maybe `void DefaultFileSource::updateOfflineMetadata(int64_t regionID, OfflineRegionMetadata, std::function<void (std::exception_ptr)>)`?

**Comment by tobrun** [comment]
Re https://github.com/mapbox/mapbox-gl-native/pull/6338#issuecomment-247694944:
Yes, this is a lot cleaner. Will fix this while providing the binding integration. 

Re https://github.com/mapbox/mapbox-gl-native/pull/6338#issuecomment-247652638:
Got some 👀 from @tmpsantos and it all makes sense now. It's not because of the move only class but due not adding `const` to the metadata when passing this into a thread invoke.

**Comment by tobrun** [comment]
I fixed up the commits and this PR is ready for review

**Pull request merged**